### PR TITLE
[DOCS] Adding get snapshot status API docs

### PR DESCRIPTION
**Pull request by lockewritesdocs**
Adds a new page for the [Get snapshot status API](https://elasticsearch_59355.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/get-snapshot-status-api.html).

Backports:

* 7.x #59670
* 7.9 #59711
* 7.8 #59712

Relates to #56069

### CODE REVIEW

**Comment by lockewritesdocs** [comment]
@jrodewig, the tests that I added are failing for both `start_time_in_millis` and `uuid`, though I can't figure out why. The build indicates that it expected one value for `uuid` [but returned another](ITWHQY2fR2OqX-TYt6mUNA). Shouldn't `[s/s/"uuid"]` substitute a value and say "don't anticipate this exact value, just match against what is in the response"?

File: docs/reference/snapshot-restore/apis/get-snapshot-status-api.asciidoc [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+        "time_in_millis" : 0
+      },
+      "indices" : { }
+    }
+  ]
+}
+----
+// TESTRESPONSE[s/"uuid": "ITWHQY2fR2OqX-TYt6mUNA"/"uuid": $body.snapshots.0.uuid/]
+// TESTRESPONSE[s/"start_time_in_millis": 1594158151205/"start_time_in_millis": $body.snapshots.0.start_time_in_millis/]
+// TESTRESPONSE[s/"time_in_millis": 0/"time_in_millis": $body.snapshots.0.time_in_millis/]
```
Comment (jrodewig) [inline docs/reference/snapshot-restore/apis/get-snapshot-status-api.asciidoc:228]: Couple of errors here:
*  The response includes a space between the property name and the `:`. However your comments don't include this space. To fix, change either so they're consistent.
* The `start_time_in_millis` and `time_in_millis` properties are nested in the `stats` property. You'll need to add this property to your respective comments.

```suggestion
// TESTRESPONSE[s/"uuid" : "ITWHQY2fR2OqX-TYt6mUNA"/"uuid" : $body.snapshots.0.uuid/]
// TESTRESPONSE[s/"start_time_in_millis" : 1594158151205/"start_time_in_millis" : $body.snapshots.0.stats.start_time_in_millis/]
// TESTRESPONSE[s/"time_in_millis" : 0/"time_in_millis" : $body.snapshots.0.stats.time_in_millis/]
```

**Review by jrodewig — COMMENTED** [review]
@lockewritesdocs. I left a few comments that should help fix the CI tests. Mostly minor errors that were preventing the substitutions.

File: docs/reference/snapshot-restore/apis/get-snapshot-status-api.asciidoc (hunk 1 above)
Comment (lockewritesdocs) [reply docs/reference/snapshot-restore/apis/get-snapshot-status-api.asciidoc:228]: Thank you @jrodewig -- I originally had `.stats` in the test but removed it because I thought that it could be breaking the CI tests. However, just missing spaces 😆

File: docs/reference/snapshot-restore/apis/get-snapshot-status-api.asciidoc [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+
+Use the get snapshot status API to retrieve detailed information about snapshots currently running in the cluster.
+
+If you specify both the repository name and snapshot, the request retrieves detailed status information for the given snapshot, even if not currently running.
+
+[[get-snapshot-status-api-path-params]]
+==== {api-path-parms-title}
+
+`<repository>`::
+(Required, string)
```
Comment (jrodewig) [inline docs/reference/snapshot-restore/apis/get-snapshot-status-api.asciidoc:48]: This should be optional based on the `Omitting...` sentence below.
```suggestion
(Optional, string)
```

File: docs/reference/snapshot-restore/apis/get-snapshot-status-api.asciidoc [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+Snapshot repository name used to limit the request.
++
+Omitting a repository name retrieves a list of all snapshots currently running in the cluster.
```
Comment (jrodewig) [inline docs/reference/snapshot-restore/apis/get-snapshot-status-api.asciidoc:51]: You also need to omit the `<snapshot>` param, right?

I'd reword this so it's imperative and places the desired info at the beginning:

`To get a list of all snapshots currently running in the cluster, omit this and the `<snapshot>` parameter.

Also: are wildcard expressions supported?

File: docs/reference/snapshot-restore/apis/get-snapshot-status-api.asciidoc [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+
+`<snapshot>`::
+(Required, string)
```
Comment (jrodewig) [inline docs/reference/snapshot-restore/apis/get-snapshot-status-api.asciidoc:54]: This param looks to be optional
```suggestion
(Optional, string)
```

Also: are wildcard expressions supported?

File: docs/reference/snapshot-restore/apis/get-snapshot-status-api.asciidoc [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+          "size_in_bytes" : 0
+        },
+        "start_time_in_millis" : 1594158151205,
+        "time_in_millis" : 0
+      },
+      "indices" : { }
+    }
+  ]
+}
+----
```
Comment (jrodewig) [inline docs/reference/snapshot-restore/apis/get-snapshot-status-api.asciidoc:263]: It'd be good to create a couple of indices and add some documents to them so this is a little more realistic.

Currently, all the stats are zero and there are no indices, which is where a good bit of the stats are located.

File: docs/reference/snapshot-restore/apis/get-snapshot-status-api.asciidoc [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+If `true`, the request ignores snapshots that are unavailable, such as those that are corrupted or temporarily cannot be returned.
+
+[role="child_attributes"]
+[[get-snapshot-status-api-response-body]]
+==== {api-response-body-title}
+
+`snapshot`::
+(string)
+Name of the snapshot.
+
```
Comment (jrodewig) [inline docs/reference/snapshot-restore/apis/get-snapshot-status-api.asciidoc:74]: Missing the `repository` property.

File: docs/reference/snapshot-restore/apis/get-snapshot-status-api.asciidoc [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+`SUCCESS`::
+  The snapshot finished and all shards were stored successfully.
+====
+--
+
+`include_global_state`::
+(boolean)
+Indicates whether the current cluster state is included in the snapshot.
+
+[[get-snapshot-status-shards-stats]]
```
Comment (jrodewig) [inline docs/reference/snapshot-restore/apis/get-snapshot-status-api.asciidoc:108]: Looks like a copy/paste error.
```suggestion
```

File: docs/reference/snapshot-restore/apis/get-snapshot-status-api.asciidoc [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+`stats`::
+(object)
+Provides details on the number (`file_count`) and size (`size_in_bytes`) of files included in the snapshot.
++
+.Properties of `stats`
+[%collapsible%open]
+====
+`incremental`::
+(object)
+Number and size of files that are not already in the repository and need to be copied as part of the incremental snapshot.
```
Comment (jrodewig) [inline docs/reference/snapshot-restore/apis/get-snapshot-status-api.asciidoc:151]: I'd nest another collapsible section here to document the `file_count` and `size_in_bytes` properties.

File: docs/reference/snapshot-restore/apis/get-snapshot-status-api.asciidoc [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+
+`processed`::
+(object)
+Number and size of files that are still being copied.
```
Comment (jrodewig) [inline docs/reference/snapshot-restore/apis/get-snapshot-status-api.asciidoc:155]: Same comment here about documenting the nested properties.

File: docs/reference/snapshot-restore/apis/get-snapshot-status-api.asciidoc [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+
+`total`::
+(object)
+Total number and size of files that are referenced by the snapshot.
```
Comment (jrodewig) [inline docs/reference/snapshot-restore/apis/get-snapshot-status-api.asciidoc:159]: Same about documenting the nested property.

File: docs/reference/snapshot-restore/apis/get-snapshot-status-api.asciidoc [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+
+`start_time_in_millis`::
+(long)
+The time, in milliseconds, when the snapshot creation process started.
+
+`time_in_millis`::
+(long)
+The total time, in milliseconds, that
```
Comment (jrodewig) [inline docs/reference/snapshot-restore/apis/get-snapshot-status-api.asciidoc:167]: We generally try to avoid using Java-specific data types.

```suggestion
`start_time_in_millis`::
(integer)
The time, in milliseconds, when the snapshot creation process started.

`time_in_millis`::
(integer)
The total time, in milliseconds, that
```

File: docs/reference/snapshot-restore/apis/get-snapshot-status-api.asciidoc [hunk 12]
```diff
@@ -0,0 +1,266 @@
...
+====
+
+`indices`::
+(list of objects)
+List of objects containing information about the
+indices included in the snapshot.
++
+.Properties of `indices`
+[%collapsible%open]
```
Comment (jrodewig) [inline docs/reference/snapshot-restore/apis/get-snapshot-status-api.asciidoc:176]: Technically these are properties of `<index>`. I'd document the `<index>` property and nest accordingly.

File: docs/reference/snapshot-restore/apis/get-snapshot-status-api.asciidoc [hunk 13]
```diff
@@ -0,0 +1,266 @@
...
+====
+`shards_stats`::
+See <<get-snapshot-status-shards-stats,`shards_stats`>>.
```
Comment (jrodewig) [inline docs/reference/snapshot-restore/apis/get-snapshot-status-api.asciidoc:179]:
```suggestion
`shards_stats`::
(object)
See <<get-snapshot-status-shards-stats,`shards_stats`>>.
```

File: docs/reference/snapshot-restore/apis/get-snapshot-status-api.asciidoc [hunk 14]
```diff
@@ -0,0 +1,266 @@
...
+
+`stats`::
```
Comment (jrodewig) [inline docs/reference/snapshot-restore/apis/get-snapshot-status-api.asciidoc:181]:
```suggestion
`stats`::
(object)
```

File: docs/reference/snapshot-restore/apis/get-snapshot-status-api.asciidoc [hunk 15]
```diff
@@ -0,0 +1,266 @@
...
+`FINALIZE`::
+Number of shards containing a copy of the snapshot in the finalizing stage.
+
+`INIT`::
+Number of shards containing a copy of the snapshot in the initializing stage.
+
+`STARTED`::
+Number of shards containing a copy of the snapshot in the started stage.
+======
+=====
```
Comment (jrodewig) [inline docs/reference/snapshot-restore/apis/get-snapshot-status-api.asciidoc:214]: Missing a few `<shard>` properties:

* `stats` and its child properties
* `total` and its child properties
* `start_time_in_millis`
* `time_in_millis`

Even though these are somewhat redundant, it's helpful for devs to know what schema to expect.

File: docs/reference/snapshot-restore/apis/get-snapshot-status-api.asciidoc [hunk 16]
```diff
@@ -0,0 +1,266 @@
...
+====
+
+[[get-snapshot-status-api-example]]
+==== {api-example-title}
+
+The following request returns information for `snapshot_2` in the `my_repository` repository.
```
Comment (jrodewig) [inline docs/reference/snapshot-restore/apis/get-snapshot-status-api.asciidoc:220]: I'd reword to differentiate between this and the get snapshot API.

**Review by jrodewig — APPROVED** [review]
Overall LGTM.

The only blocking suggestions are that the `<repository>` and `<snapshot>` params look to be optional.

Feel free to ignore my other comments as you see fit. Thanks!

File: docs/reference/snapshot-restore/apis/get-snapshot-status-api.asciidoc (hunk 6 above)
Comment (lockewritesdocs) [reply docs/reference/snapshot-restore/apis/get-snapshot-status-api.asciidoc:74]: Added that property and a description:
```
`repository`::
(string)
Name of the repository that includes the snapshot.
```

File: docs/reference/snapshot-restore/apis/get-snapshot-status-api.asciidoc (hunk 3 above)
Comment (lockewritesdocs) [reply docs/reference/snapshot-restore/apis/get-snapshot-status-api.asciidoc:51]: To retrieve a list of all snapshots currently running in the cluster, you omit both `repository` and `snapshot`. To get a list of all snapshots running in a repository, you only omit `snapshot`. Also, wildcards are supported for `repository` (in 8.0) but not for `snapshot`. Therefore, I made some changes to both parameters:

```
`<repository>`::
(Required, string)
Snapshot repository name used to limit the request. Wildcard (`*`) expressions are supported.
+
To retrieve a list of all snapshots in the cluster, omit this and the `<snapshot>` parameter.

`<snapshot>`::
(Required, string)
Comma-separated list of snapshot names to retrieve status for.
+
To retrieve a list of all snapshots in a specified repository, omit this parameter.
+
NOTE: Wildcard (`*`) expressions are not supported for `<snapshot>`.
```

File: docs/reference/snapshot-restore/apis/get-snapshot-status-api.asciidoc [hunk 17]
```diff
@@ -0,0 +1,273 @@
...
+.Properties of `stats`
+[%collapsible%open]
+====
+`incremental`::
+(object)
+Number and size of files that are not already in the repository and need to be copied as part of the incremental snapshot.
+
+`processed`::
+(object)
+Number and size of files that are still being copied.
```
Comment (original-brownbear) [inline docs/reference/snapshot-restore/apis/get-snapshot-status-api.asciidoc:162]: This is not correct. `processed` tracks those files that have already been uploaded. We increment the stats here after we upload a file

File: docs/reference/snapshot-restore/apis/get-snapshot-status-api.asciidoc [hunk 18]
```diff
@@ -0,0 +1,273 @@
...
+`stats`::
+(object)
+Provides details on the number (`file_count`) and size (`size_in_bytes`) of files included in the snapshot.
++
+.Properties of `stats`
+[%collapsible%open]
+====
+`incremental`::
+(object)
+Number and size of files that are not already in the repository and need to be copied as part of the incremental snapshot.
```
Comment (original-brownbear) [inline docs/reference/snapshot-restore/apis/get-snapshot-status-api.asciidoc:158]: This needs to be worded a little differently. If you run this API for an already complete snapshot then it would be 

>  Number and size of files that were not already in the repository and needed to be copied as part of the incremental snapshot. 

I think we should explain both cases here in the wording somehow.

File: docs/reference/snapshot-restore/apis/get-snapshot-status-api.asciidoc [hunk 19]
```diff
@@ -0,0 +1,273 @@
...
+=====
+`stage`::
+(string)
+Indicates the current state of the shards that include the snapshot.
++
+.Properties of `stage`
+[%collapsible%open]
+======
+`DONE`::
+Number of shards containing a completed copy of the snapshot.
```
Comment (original-brownbear) [inline docs/reference/snapshot-restore/apis/get-snapshot-status-api.asciidoc:207]: This is confusing (same for the other stage explanations here). i think something along the lines of the following would be better:

> Number of shards in the snapshot that were successfully stored in the repository.

**Review by original-brownbear — CHANGES_REQUESTED** [review]
Thanks @lockewritesdocs looks good overall, just a few points on the details.

Also, could we add a disclaimer/warning to this API that makes it clear that this might be very expensive to execute if you run it over a large number of snapshots or over snapshots that contain a large number of shards? People should really not be using this API as part of e.g. orchestration logic unless absolutely necessary IMO.

File: docs/reference/snapshot-restore/apis/get-snapshot-status-api.asciidoc (hunk 17 above)
Comment (lockewritesdocs) [reply docs/reference/snapshot-restore/apis/get-snapshot-status-api.asciidoc:162]: Thanks for the correction. I'm proposing this text instead, which will also be reflected in [Restoring a snapshot](https://github.com/elastic/elasticsearch/blame/master/docs/reference/snapshot-restore/restore-snapshot.asciidoc#L258):

>Number and size of files that have already been uploaded to the snapshot. After a file is uploaded, the total `file_count` and `size_in_bytes` are incremented in `stats`.

Comment (original-brownbear) [reply docs/reference/snapshot-restore/apis/get-snapshot-status-api.asciidoc:162]: > After a file is uploaded, the total file_count and size_in_bytes are incremented in stats.

This is actually not correct. The total is set when the shard snapshot starts. It never changes during the snapshot. Both total and incremental counts are set at the beginning and the processed is then incremented as the snapshot executes.

Comment (lockewritesdocs) [reply docs/reference/snapshot-restore/apis/get-snapshot-status-api.asciidoc:162]: Ah, so the _processed_ `file_count` and `size_in_bytes` are incremented, not the _total.

File: docs/reference/snapshot-restore/apis/get-snapshot-status-api.asciidoc (hunk 18 above)
Comment (lockewritesdocs) [reply docs/reference/snapshot-restore/apis/get-snapshot-status-api.asciidoc:158]: I think that we can amend the original statement to focus on the files that still need to be copied, versus the delta of files that didn't already exist and need copying:

>Number and size of files that still need to be copied as part of the incremental snapshot.

>For completed snapshots, this property indicates the number and size 
>of files that were not already in the repository and were copied as 
>part of the incremental snapshot.

File: docs/reference/snapshot-restore/apis/get-snapshot-status-api.asciidoc (hunk 5 above)
Comment (lockewritesdocs) [reply docs/reference/snapshot-restore/apis/get-snapshot-status-api.asciidoc:270]: Good call 👍 I'm adding a few indices locally to generate tangible figures for those properties. I'll update the examples with that data.

**Review by original-brownbear — APPROVED** [review]
LGTM thanks for the extra iterations @lockewritesdocs !
